runtime/src/accounts_db.rs

```diff
@@ -1464,39 +1464,42 @@ impl AccountsDb {
         }
     }
 
-    fn collect_uncleaned_pubkeys_to_slot(&self, max_slot: Slot) -> (Vec<Vec<Pubkey>>, Slot) {
-        let mut max_slot_in_uncleaned_pubkeys = 0;
-        let slots: Vec<Slot> = self
-            .uncleaned_pubkeys
+    /// Collect all the uncleaned slots, up to a max slot
+    ///
+    /// Search through the uncleaned Pubkeys and return all the slots, up to a maximum slot.
+    fn collect_uncleaned_slots_up_to_slot(&self, max_slot: Slot) -> Vec<Slot> {
+        self.uncleaned_pubkeys
             .iter()
             .filter_map(|entry| {
-                let slot = entry.key();
-                max_slot_in_uncleaned_pubkeys = max_slot_in_uncleaned_pubkeys.max(*slot);
-                if *slot <= max_slot {
-                    Some(*slot)
-                } else {
-                    None
-                }
+                let slot = *entry.key();
+                (slot <= max_slot).then(|| slot)
             })
-            .collect();
-        (
-            slots
-                .into_iter()
-                .filter_map(|slot| {
-                    let maybe_slot_keys = self.uncleaned_pubkeys.remove(&slot);
-                    if self.accounts_index.is_root(slot) {
-                        // Safe to unwrap on rooted slots since this is called from clean_accounts
-                        // and only clean_accounts operates on rooted slots. purge_slots only
-                        // operates on uncleaned_pubkeys
```

Comment:
You'll notice that I took out the check for `is_root()`. Shouldn't _all_ the pubkeys be collected from slot removal, instead of just the pubkeys that came from rooted slots?

I was thinking that either (1) all the slots in `uncleaned_pubkeys` are rooted slots anyway, or (2) if there are _non-rooted_ slots, then those pubkeys should definitely be cleaned up.

Am I missing something here?

Reply:
@brooksprumo (Okay, I was forgetting everything about AccountsDb and remembered it to some extent..)

you have a good questions from the newcomer's eyes. :) And, I think this is good exercise for you to read through the AccountsDb code.

> (1) all the slots in uncleaned_pubkeys are rooted slots anyway

I don't think this is true. And I think this is easy to prove. With that hint given, could you read the code to find the _why_?

> (2) if there are non-rooted slots, then those pubkeys should definitely be cleaned up.

as you say, the need of cleaning is right. but non-rooted slots should be handled carefully. and that's fun part of AccountsDB. Unless rooted, banks are like snapshots/immutable collection for the whole accounts state. So, each bank must maintain delta of account updates. That's why we pass around `Ancestors`. So, new minor fork (child bank) can be created from any banks (parent banks). And then those un-rooted banks are finally promoted to rooted banks by  `squash()` and its caller. I think these code are also helpful to understand the high-level workings.

sorry for a bit being teasing. no intention to be mean to you. I just want you to be levelled up for AccountsDB. Code reading to fill the questions with hints might be good excises for you. :)

Reply:
@ryoqun Yes, these will included rooted slots, but because we only call `clean_acounts` on a rooted bank and/or we only iterate up to a rooted slot, these slots will be unrooted slots below the root and thus are dead and cleanable. Is there any issue you see?

We call `clean_accounts` on rooted banks:
https://github.com/solana-labs/solana/blob/8f56c116d71e5349b8a63e5d478c37033cfbdc2c/runtime/src/accounts_background_service.rs#L305

Max slot here will either be highest root or root -1:
https://github.com/solana-labs/solana/blob/8f56c116d71e5349b8a63e5d478c37033cfbdc2c/runtime/src/bank.rs#L4720

```
                // Note this only cleans up the storage entries. The accounts index cleaning
                // (removing from the slot list, decrementing the account ref count), is handled in
                // clean_accounts() -> purge_older_root_entries()
```
https://github.com/solana-labs/solana/blob/8f56c116d71e5349b8a63e5d478c37033cfbdc2c/runtime/src/accounts_db.rs#L3034-L3036

`clean_accounts -> purge_older_root_entries` won't be operating on the keys touched in these unrooted slots if it does not look at the list of delta keys.

Reply:
@sakridge yeah, it looks like you're right. those `slots` are actually `roots` proven by the current codepath. I was short-sighted...

@brooksprumo sorry for confusion. So, it looks like `if self.accounts_index.is_root(slot) {` has been redundant check to begin with. Then removing this check sounds ok to me according to @sakridge 's reasonsing.

> > (1) all the slots in `uncleaned_pubkeys` are rooted slots anyway

> I don't think this is true.

Here's my thinking regarding the above statement:

(I failed once but if I read the code correctly,) `uncleaned_pubkeys` itself should be mixed with rooted and unrooted slots at the time of cleaning because the below is the sole code `insert()`-ing into `uncleaned_pubkeys` and it's called for both the cases of rooted and unrooted at the time of the insertion (immediately before `Bank.freeze()`):

https://github.com/solana-labs/solana/blob/1c95e2bbee0a221f7ff4a4666605363aabc26a74/runtime/src/accounts_db.rs#L4269

Does this code reading align with your reading?

Reply:
@ryoqun 
>> I don't think this is true.

> Here's my thinking regarding the above statement:

yes, I believe this statement is correct. The slots in `uncleaned_pubkeys` will be all rooted and unrooted slots.

I believe the origination of this is that previously the `clean_accounts` only operated on rooted slots, so we kept that behavior. Although the caching code may have added new assumptions about where unrooted slots were being cleaned up in the index.